packages/x-components/src/x-modules/url/components/url-handler.vue

```diff
@@ -1,82 +1,209 @@
-      this.saveUrlConfig();
+      window.addEventListener('load', this.emitEvents.bind(this));
+      window.addEventListener('popstate', this.emitEvents.bind(this));
+    }
 
-      window.addEventListener('load', () => {
-        this.$x.emit('DocumentLoaded', window.location.href);
-        this.emitOpenEvent();
-      });
-      window.addEventListener('popstate', () => {
-        this.$x.emit('DocumentLoaded', window.location.href);
-        this.emitOpenEvent();
-      });
+    beforeDestroy(): void {
+      window.removeEventListener('load', this.emitEvents.bind(this));
+      window.removeEventListener('popstate', this.emitEvents.bind(this));
+    }
+
+    /**
+     * Updates the browser URL with the new {@link UrlParams} using the history `pushState` method.
+     *
+     * @param newUrlParams - The new params to update browser URL.
+     */
+    @XOn('PushableUrlStateChanged')
+    updateUrlWithPush(newUrlParams: UrlParams): void {
+      this.updateUrl(newUrlParams, window.history.pushState.bind(window.history));
+    }
+
+    /**
+     * Updates the browser URL with the new {@link UrlParams} using the history `replaceState`
+     * method.
+     *
+     * @param newUrlParams - The new params to update browser URL.
+     */
+    @XOn('ReplaceableUrlStateChanged')
+    updateUrlWithReplace(newUrlParams: UrlParams): void {
+      this.updateUrl(newUrlParams, window.history.replaceState.bind(window.history));
     }
 
     /**
-     * If present, save the url parameters names in the store passed as prop and read as attrs.
+     * Emits the {@link XEventsTypes.ParamsLoadedFromUrl} XEvent and, if there is query, also emits
+     * the  {@link XEventsTypes.UserOpenXProgrammatically}.
      *
      * @internal
      */
-    protected saveUrlConfig(): void {
-      if (Object.keys(this.$attrs).length > 0) {
-        const params = reduce(
-          this.$attrs,
-          (urlParamsConfig, paramName, urlParamName) => {
-            urlParamsConfig[paramName] = urlParamName;
-            return urlParamsConfig;
-          },
-          {} as Record<keyof UrlConfig['urlParamNames'], string>
-        );
-
-        this.$x.emit('UrlConfigProvided', {
-          urlParamNames: {
-            ...params
-          }
-        });
+    protected emitEvents(): void {
+      const urlParams = this.getUrlParams();
+      this.$x.emit('ParamsLoadedFromUrl', urlParams);
+      if (urlParams.query) {
+        this.$x.emit('UserOpenXProgrammatically');
+      }
+    }
+
+    /**
+     * Gets the {@link UrlParams} from the URL, including only the params defined by
+     * {@link paramsNames}.
+     *
+     * @returns UrlParams obtained from URL.
+     * @internal
+     */
+    protected getUrlParams(): UrlParams {
+      const urlSearchParams = new URL(window.location.href).searchParams;
+      const urlParams = this.paramsNames.reduce((urlParams, paramName) => {
+        urlParams[paramName] = this.getParamByType(urlSearchParams, paramName);
```

Comment:
I buy you the `if` with the `has` but the other changes...

You are simplifying a method to complicate the other. I prefer to leave in one single method the responsibility of getting one URL param

Reply:
also is not a good practice to cast types with `!` if not necessary. TypeScript is here to help, if you silence him, then it makes no sense.

Reply:
I know and you are right but in this particular case, once you know `urlSearchParams.has(param)` is `true` then you are 100% certain `urlSearchParams.get(param)` will not be null. But Typescript doesn't know that.

Actually now I realise the return of the boolean `case` is also wrong. The `?` is opening the possibility to return `undefined`, but the function is not typed to return undefined. I would use `!` here also because again, you are 100% sure `urlSearchParams.get(param)` will return a value (a string actually)

Reply:
Another doubt I have. Is there a preference to use?

`return Object.assign({}, initialUrlState.params, urlParams);`
 instead of 
`return {...initialUrlState.params, ...urlParams}`